src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -546,6 +592,127 @@ case object LOG3 extends LogOp(0xa3)
 case object LOG4 extends LogOp(0xa4)
 
 
+sealed abstract class CallOp(code: Int, delta: Int, alpha: Int) extends OpCode(code, delta, 1, G_zero) {
+
+  protected def exec(state: ProgramState): ProgramState = {
+    val (Seq(gas, to, callValue, inAddr, inSize, outAddr, outSize), stack1) = getParams(state)
+
+    val (inputData, mem1) = state.memory.load(inAddr, inSize)
+    val endowment = if (this == DELEGATECALL) BigInt(0) else callValue.toBigInt
+
+    val startGas = {
+      val gExtra = gasExtra(state, endowment, Address(to))
+      val gCap = gasCap(state, gas, gExtra)
+      if (endowment == 0) gCap else gCap + G_callstipend
+    }
+
+    lazy val result = {
+      val toAddr = Address(to)
+
+      val (account, world1) = this match {
+        case CALL =>
+          val w = state.world.transfer(state.ownAddress, toAddr, endowment)
+          val acc = w.getGuaranteedAccount(toAddr)
+          (acc, w)
+
+        case _ =>
+          val acc = state.world.getAccount(toAddr).getOrElse(Account.Empty)
+          (acc, state.world)
+      }
+
+      val (owner, caller) = this match {
+        case CALL => (toAddr, state.ownAddress)
+        case CALLCODE => (state.ownAddress, state.ownAddress)
+        case DELEGATECALL => (state.ownAddress, state.env.callerAddr)
+      }
+
+      val env = state.env.copy(
+        ownerAddr = owner,
+        callerAddr = caller,
+        inputData = inputData,
+        value = callValue,
+        program = Program(world1.getCode(account.codeHash)),
+        callDepth = state.env.callDepth + 1)
+
+      val context = state.context.copy(
+        env = env,
+        startGas = startGas,
+        world = world1)
+
+      VM.run(context)
+    }
+
+    val validCall =
+      state.env.callDepth < OpCode.MaxCallDepth &&
+      endowment <= state.ownAccount.balance
+
+    if (!validCall || result.error.isDefined) {
+      val stack2 = stack1.push(DataWord.Zero)
+      val gasAdjustment = if (validCall) BigInt(0) else -startGas
+
+      state
+        .withStack(stack2)
+        .withMemory(mem1)
+        .spendGas(gasAdjustment)
+        .step()
+
+    } else {
+      val stack2 = stack1.push(DataWord(1))
+      val output = result.returnData.take(outSize.intValue)
+      val mem2 = mem1.store(outAddr, result.returnData)
```

Comment:
You don't need to remember, you need to go that line and see what happened there. This would be the same `stackAfterCall` which is almost meaningless - what call, how was it changed on that call? 

Here's a descriptive name: `val stackWithTheResultOfSuccessfulExternalCodeExecution =`. Is that something you'd really like to see? 😁 Actually you still don't know from the name what value was pushed onto the stack 😉 

Using this pseudo-mathematical notation it's at least clear `state_(n+1)` should result from modification of `state_n`, and that the final state is the one with highest `n`. That cannot be said for those "descriptive" names.